rasa/cdu/command_processor.py

```diff
@@ -174,21 +184,29 @@ def execute_commands(
                     "command_executor.skip_cancel_flow.no_active_flow", command=command
                 )
                 continue
-            # in between the prediction and this canceling command, we might have
-            # added some stack frames. hence, we can't just cancel the current top frame
-            # but need to find the frame that was at the top before we started
-            # processing the commands.
-            for idx, frame in enumerate(flow_stack.frames):
-                if frame.flow_id == current_top_flow.id:
-                    structlogger.debug("command_executor.cancel_flow", command=command)
-                    # Setting the flow to the end step so it is properly
-                    # wrapped up by the flow policy
-                    flow_stack.frames[idx].step_id = END_STEP
-            events.append(SlotSet(CANCELLED_FLOW_SLOT, current_top_flow.id))
+
+            canceled_frames = []
+            original_frames = FlowStack.from_dict(original_stack_dump).frames
+            # we need to go through the original stack dump in reverse order
+            # to find the frames that were canceled. we cancel everthing from
+            # the top of the stack until we hit the user flow that was canceled.
+            # this will also cancel any patterns put ontop of that user flow,
+            # e.g. corrections.
```

Comment:
We should probably test what the llm does when someone tries to cancel a correction by saying "no, stop" or "cancel". Could be an e2e test case.

Reply:
oh yeah, I think you are handling the "don't do the correction case". I am just wondering about the interplay between "no" and "cancel" here. If someone says cancel to the question "Do you want to update the recipient?" then they might end up cancelling the whole transfer money flow, as cancellations dig down to the user frame